### ISSUE TITLE
Convex geometry clean up

### ISSUE THREAD

**Issue by SeanCurtis-TRI** [flexible-collision-library/fcl#326]
Perusing the [`Convex`](https://github.com/flexible-collision-library/fcl/blob/master/include/fcl/geometry/shape/convex.h) geometry implementation, several issues jump out:

1. The class has several unused members. They aren't used in the class's internal implementation and, it likewise appears, are not directly referenced (as public members). They should be removed from the class API.
    - `plane_normals`
    - `plane_dis`
    - `edges`
2. The class does no validation on the inputs. It can either validate by default at construction or be available as a user-callable a method such that the user can choose to validate.
3. The operations that calculate volume, center of mass, and moment of inertia tensor all have duplicated geometry (they all rely on the volume-weighted combination of smaller tetrahedra). To iterate through the tetrahedra and compute volume is *duplicated* across all methods. Refactor so it's only defined once.
4. Currently, the smaller tetrahedra are defined generally (every face is divided into tetrahedra based on the face's "center" point and a fan of triangles including all other vertices). If the face is actually a triangle, this leads to volume for three tetrahedra being computed instead of just the one. The decomposition of the geometry into tetrahedra should, at least, handle triangular faces more intelligently. And perhaps also 4-sided faces.
5. There are no (until #325 and then limited) unit tests for the geometry.

**Comment by SeanCurtis-TRI** [comment]
@jslee02 Do you know any reason why I can't pull the trigger on modifying the API of the `Convex` type?

**Comment by jslee02** [comment]
@SeanCurtis-TRI Nope. I think your points are all valid. Looking at the history, `Convex` class was added in the very beginning of FCL (7 years ago), and there hasn't been any significant improvement made. It's time to move on!

As a note, one of my personal reasons I haven't taken a look at this class is that I believed there is no performance advantage of using `Convex` over the BVH representation even for convex meshes. This is because FCL uses the convexity-based algorithms for `Convex`, and [the specific algorithm frequently requires iteration over all the vertices](https://github.com/flexible-collision-library/fcl/blob/master/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L2106). This conclusion is made without a comprehensive benchmark so it could be wrong. However, the current algorithm can be replaced by a more efficient algorithm anytime so I think this modification worth enough.

**Comment by SeanCurtis-TRI** [comment]
Hi @jslee02 Thanks for weighing in.

I know what you mean about the support function for `Convex`. In its current incarnation, it does a linear pass over all vertices, period. However, that *can* be sped up and, I believe, probably should be. Here are the things to consider.

1. The one thing the `Convex` class gives us (in conjunction with GJK/MPR) is that it will detect if one convex shape is completely contained within another. Pushing it into a BVH as a mesh will only detect collision if there's actual intersection at the surface.
2. The support function can be made a lot faster by making use of the topology (i.e., the edges). Rather than a linear search, the code can basically walk along the surface of the mesh based on edge directions. That would certainly reduce the linear component. I suspect that was *probably* the motivation for the edges in the first place. (Or, at the very least, I'm hard-pressed thinking up an alternative motivation.)

But that's for the future. In the short-term, we'll go ahead and clean it up so it's, at least, a fully participatory member of FCL.

**Comment by SeanCurtis-TRI** [comment]
@jslee02 Quick question, I've added one more point above (about data ownership). The `Convex` class doesn't own the data it is provided; it's simply given pointers. Do you have any insight into that design decision? While I can imagine a case where the `Convex` class could simply be an abstraction on top of some other data, I would imagine the default case is that it is a stand-alone entity with its own memory and self-contained lifespan. Yes?

**Comment by SeanCurtis-TRI** [comment]
Great. A shared_ptr for the data would definitely resolve those issues. That would make it so that no external entity has to keep track of the `Convex`'s geometry but would allow the same geometry to be used by multiple `Convex` instances. And I was thinking of a `std::shared_ptr<std::vector<Thing>>` just to modernize things a bit.

**Comment by MengeCrowdSim** [comment]
If you've got an example that exposes some of the new asserts, we would love to get that into a unit test. We've had some issues reported in the past, but have been unable to reproduce the issues (which makes debugging nigh on to impossible). So, if you can do that as a PR, that'd be brilliant. And the sooner the better -- I'd love if we can get that in before the next release.

We'll have a chat about qhull dependency. Clearly the FCL API was designed with minimal dependencies -- clearly they didn't want to have to compute a convex hull and relied on one being pre-computed. A method that facilitates that would definitely be a *huge* improvement in the usability of the API. The only question in my mind comes down to making sure the dependency "makes sense" (whatever vague concepts that simple phrase captures). We've discussed it in the past and bringing qhull probably makes sense. If we pull the trigger on that, it'd make sense to defer that until after we release the next public release.

As for the error you're finding in your hulls, when you say the answer is off, can you elaborate?  What are the radii of the spheres? Are the tesselations polar or geodesic? What are the sizes of the triangles/faces? Is that error compared to the idealized spheres or the known tesselation? There had been an earlier issue related to that, and the answers that were being produced were exact, to the resolution of the mesh. I want to make sure this isn't another instance of that (although 12 cm is pretty large -- unless, of course, the spheres are much, much larger).

**Comment by SeanCurtis-TRI** [comment]
Oops -- that last comment was mine. That's what happens when I respond at home -- I forget I'm not logged in as myself.  :-)

**Comment by Levi-Armstrong** [comment]
I will create a few unit tests that cause the assert failures. Also for these tests instead of relying on qhull I will write the data to a simple text file which gets parsed by the unit test.

**Comment by SeanCurtis-TRI** [comment]
You rock! That'll be great. One of the things I'm hoping is that the test will still fail if we take your tesselated sphere and throw out most of the faces away from the contact. (the fewer faces there are, the easier it is to debug -- but with GJK/EPA algorithms, the path you take through the code is sensitive to the actual feature set you have.) So, we'll start with the data that actually reproduces the problem and only try simplifying it to the extent that it still works. So, you don't have to waste your time tinkering with "simpler" geometry.
